lib/trino-hive-formats/src/main/java/io/trino/hive/formats/line/json/JsonDeserializer.java

```diff
@@ -0,0 +1,838 @@
+                throws IOException
+        {
+            if (parser.currentToken() != START_OBJECT) {
+                throw invalidJson("start of object expected");
+            }
+
+            int[] jsonToRowIndex = getJsonToRowIndex(lineBuffer, parser);
+            boolean[] fieldWritten = new boolean[fieldNames.size()];
+
+            int jsonFieldIndex = 0;
+            while (nextObjectField(parser)) {
+                int rowIndex = jsonToRowIndex[jsonFieldIndex];
+                jsonFieldIndex++;
+                if (rowIndex < 0) {
+                    skipNextValue(parser);
+                    continue;
+                }
+
+                nextTokenRequired(parser);
+                fieldDecoders.get(rowIndex).decode(lineBuffer, parser, fieldBuilders.apply(rowIndex));
+                fieldWritten[rowIndex] = true;
+            }
+
+            // write null to unset fields
+            for (int i = 0; i < fieldWritten.length; i++) {
+                if (!fieldWritten[i]) {
+                    fieldBuilders.apply(i).appendNull();
+                }
+            }
+        }
+
+        private int[] getJsonToRowIndex(LineBuffer lineBuffer, JsonParser parser)
+                throws IOException
+        {
+            // create a new parser starting at the beginning of the json object
+            JsonParser fieldNameParser = createParserAt(parser.currentTokenLocation(), jsonFactory, lineBuffer);
+            if (fieldNameParser.nextToken() != START_OBJECT) {
+                throw invalidJson("start of object expected");
+            }
+
+            // build a mapping from field in the row to the field in the json object
+            int[] rowToJson = new int[fieldNames.size()];
+            Arrays.fill(rowToJson, -1);
+
+            int jsonFieldIndex = 0;
+            while (nextObjectField(fieldNameParser)) {
+                String fieldName = fieldNameParser.getText();
+                int rowIndex = getFieldPosition(fieldName);
+                skipNextValue(fieldNameParser);
+
+                if (rowIndex >= 0) {
+                    rowToJson[rowIndex] = jsonFieldIndex;
+                }
+                jsonFieldIndex++;
+            }
+
+            // reverse the mapping
+            int[] jsonToRowIndex = new int[jsonFieldIndex];
+            Arrays.fill(jsonToRowIndex, -1);
+            for (int rowIndex = 0; rowIndex < rowToJson.length; rowIndex++) {
+                int jsonIndex = rowToJson[rowIndex];
+                if (jsonIndex >= 0) {
+                    jsonToRowIndex[jsonIndex] = rowIndex;
+                }
+            }
+            return jsonToRowIndex;
+        }
+
+        private int getFieldPosition(String fieldName)
+        {
+            int fieldPosition = fieldNames.indexOf(fieldName.toLowerCase(Locale.ROOT));
+            if (fieldPosition >= 0) {
+                return fieldPosition;
+            }
+
+            // TODO: columns named _colXX are positional which is weird
```

Comment:
What needs to be done here?

Reply:
Nothing the comment should have been removed